### PR TITLE
Ignore weak HElem at the beginning and end of an LR group

### PR DESCRIPTION
**Pull request by Leedehai**
Fixing #2515.

Demo:
```typ
$ integral_S f(x) mu(dif x), norm(dif x) $
```

Before this change
Screenshot 2023-12-14 at 00 41 38

After this change
Screenshot 2023-12-14 at 00 39 32

### CODE REVIEW

**Comment by Leedehai** [comment]
Another alternative is to implement `Behave` for `TextElem` whose content is a delimiter or fence, but that carries the  downside of context ignorance, e.g. `|dif x|` doesn't need space before "d", but `|f(x)| dif x` needs a space before "d". Also, touching `TextElem` would impact text elements outside math block, and that is almost unintended.

Yet another alternative is not prepending an `HElem` before `dif` in op.rs, but prepending Math's own special space element that will be ignored if that space is at the beginning/end of an LR group, similar to this PR. `HElem` and `Behave` are not for math, anyway -- they were newly introduced to math in PR 1633.

**Comment by Leedehai** [comment]
A. An alternative is to implement `Behave` for `TextElem` whose content is a delimiter or fence (let it return `Behaviour::Destructive` instead of `Supportive`), but this idea carries downsides:
* `Behaviour::Destructive` destructs weak spaces before _and_ after the `TextElem`, not just on one side.
* LR context ignorance, since `BehavedBuilder` isn't aware of LR context. E.g. `|dif x|` actually doesn't need space before "d", but `|f(x)| dif x` actually needs a space before "d".
* Touching `TextElem` would impact text elements outside math block, and that is almost unintended.

B. Another alternative is not prepending an `HElem` before `dif` in op.rs, but prepending Math's own special weak space element that, besides behaving the same as a weak HElem, will be ignored if that space is at the beginning/end of an LR group, similar to this PR. But that is basically duplicating what `BehavedBuilder` does (in addition to the said LR thing).

C. Yet another alternative is to utilize `op(upright(d) x)` (not `op(upright(d)) x` since we don't want a space between "d" and "x") (credit to Eric), but this requires making `dif` a function instead of a symbol, and that appears to be quite a big breaking change, since it is by no means a little-used symbol, and would break an otherwise very convenient symbol. And there's also `sym.diff` that probably also needs to change for the sake of symmetry.

D. Yet again another alternative is to utilize `op(upright(d))zwj` and add a match rule in spacing.rs's `spacing()` so that somehow `(_, zwj)` and `(zwj, _)` don't yield a THIN space -- after all, that is what "zwj" (zero-width joiner) means. However, a high order differential `dif^2` would lose its preceding weak space, since the expression would be wrapped in an `AttachElem`.

**Comment by sitandr** [comment]
Just to make sure, the fact it doesn't work the way weak spacing was introduced in #1663 is not some minor later-introduced bug, but something more serious?

I haven't been really following this problem, so maybe the code changed dramatically and it has nothing common with #1663 (or #1966 partly fixing it). Just want to make sure.

**Comment by laurmaedje** [comment]
As said on Discord, I wonder if we can fix the problem more generally by not using weak h spacing, but rather changing the class of the `dif` symbol.

File: crates/typst/src/math/lr.rs [hunk 1]
```diff
@@ -77,6 +79,21 @@ impl LayoutMath for LrElem {
...
+            count += 1;
+            if count == 2 || count == original_len - 1 {
+                !matches!(
+                    fragment,
+                    MathFragment::Spacing(SpacingFragment { weak: true, .. })
+                )
+            } else {
+                true
+            }
+        });
```
Comment (laurmaedje) [inline crates/typst/src/math/lr.rs:95]:
```suggestion
        let original_len = fragments.len();
        let mut index = 0;
        fragments.retain(|fragment| {
            index += 1;
            (index != 2 && index + 1 != original_len) || !matches!(
                fragment,
                MathFragment::Spacing(SpacingFragment { weak: true, .. })
            )
        });
```

**Review by laurmaedje — COMMENTED** [review]
I guess this makes sense independently of whether dif is best implemented with weak h spacing. So I think we can go ahead with it.

**Comment by Leedehai** [comment]
Thanks. I uploaded a new patch.

**Pull request merged**